compiler-core/src/language_server/annotation_printer.rs

```diff
@@ -0,0 +1,218 @@
+            type_aliases,
+            module_aliases,
+            type_parameters,
+            current_module,
+        }
+    }
+
+    pub fn print(&mut self, typ: &Type) -> String {
+        let mut typ_str = String::new();
+        match typ {
+            Type::Named {
+                name, args, module, ..
+            } => {
+                let key = (module.clone(), name.clone());
+
+                if let Some(typ_alias) = self.type_aliases.get(&key) {
+                    // Type has been aliased. eg: import mod1.{type Cat as c} or type UserId = Int
+                    if self.name_clashes_if_unqualified(typ_alias, module) {
+                        typ_str.push_str(module);
+                        typ_str.push('.');
+                    }
+                    typ_str.push_str(typ_alias.as_str());
+                } else if module == PRELUDE_MODULE_NAME || *module == self.current_module {
+                    if self.name_clashes_if_unqualified(name, module) {
+                        typ_str.push_str(module);
+                        typ_str.push('.');
+                    }
+                    typ_str.push_str(name.as_str());
+                } else {
+                    // Always qualify types from other modules
+                    typ_str.push_str(self.module_aliases.get(module).unwrap_or(module).as_str());
+                    typ_str.push('.');
+                    typ_str.push_str(name.as_str());
+                }
+
+                if !args.is_empty() {
+                    typ_str.push('(');
+                    typ_str.push_str(&self.args_to_string(args));
+                    typ_str.push(')');
+                }
+                let _ = self
+                    .printed_types
+                    .insert(EcoString::from(typ_str.clone()), module.clone());
+                typ_str
+            }
+            Type::Fn { args, retrn } => {
+                typ_str.push_str("fn(");
+                typ_str.push_str(&self.args_to_string(args));
+                typ_str.push_str(") -> ");
+                typ_str.push_str(&self.print(retrn));
+
+                typ_str
+            }
+            Type::Var { type_: typ, .. } => match *typ.borrow() {
+                TypeVar::Link { type_: ref typ, .. } => self.print(typ),
+                TypeVar::Unbound { id, .. } | TypeVar::Generic { id, .. } => {
+                    if let Some(crate::ast::TypeAst::Var(type_var)) = self.type_parameters.get(&id)
+                    {
+                        return type_var.name.to_string();
+                    }
+                    self.generic_type_var(id)
+                }
+            },
+            Type::Tuple { elems, .. } => {
+                typ_str.push_str("#(");
+                typ_str.push_str(&self.args_to_string(elems));
+                typ_str.push(')');
+
+                typ_str
+            }
+        }
+    }
+
+    fn name_clashes_if_unqualified(&mut self, type_: &EcoString, module: &str) -> bool {
```

Comment:
What does this function mean? I'm not sure what it is doing looking at the previously printed types.

I think it's here to work out if a prelude type has been shadowed by a type defined in this module, but you can't work that out by looking at what has been printed. Rather, we need to know the names of all the types that are in scope in this module, both defined and imported.